{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Step 5. Plot AS variants and motif enrichment results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Add detailed description here "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"data.table\")\n",
    "library(ggplot2)\n",
    "library(RcisTarget)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifName <- readLines(\"/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/MM057/alleleseq/motif_analysis/singletons_names.txt\")\n",
    "# 22936\n",
    "motifName <- motifName[-1]\n",
    "motifName <- sapply(motifName, strsplit, \"\\t\")\n",
    "motifName <- setNames(sapply(motifName, function(x) x[2]), sapply(motifName, function(x) x[1]))\n",
    "\n",
    "readNformat <- function(matFile)\n",
    "{\n",
    "  mat <- fread(matFile, sep=\"\\t\", header = TRUE)\n",
    "  mat <- data.frame(mat, check.names=FALSE)\n",
    "  row.names(mat) <- mat[,1]\n",
    "  mat <- as.matrix(mat[,-1])\n",
    "  if(any(is.na(colnames(mat))))  warning(\"missing motif names\")\n",
    "  return(mat)\n",
    "}\n",
    "ASB_MUT <- readNformat(\"ASB_MUT_sorted.new_cbust_result.tsv\")\n",
    "ASB_REF <- readNformat(\"ASB_REF_sorted.new_cbust_result.tsv\")\n",
    "dim(ASB_MUT)\n",
    "dim(ASB_REF)\n",
    "ASB_delta_comb<-matrix(nrow=0,ncol=22937)\n",
    "\n",
    "ASB_MUT_losses<-ASB_MUT[grep(\"LOSES\",rownames(ASB_MUT)),]\n",
    "ASB_MUT_wins<-ASB_MUT[grep(\"WINS\",rownames(ASB_MUT)),]\n",
    "ASB_REF_losses<-ASB_REF[rownames(ASB_MUT_losses),]\n",
    "ASB_REF_wins<-ASB_REF[rownames(ASB_MUT_wins),]\n",
    "ASB_delta <- ASB_MUT_wins - ASB_REF_wins\n",
    "ASB_delta_comb<-rbind(ASB_delta_comb,ASB_delta)\n",
    "ASB_delta <- ASB_REF_losses - ASB_MUT_losses\n",
    "ASB_delta_comb<-rbind(ASB_delta_comb,ASB_delta)\n",
    "ASB_delta<-ASB_delta_comb\n",
    "saveRDS(ASB_delta, \"ASB_delta.Rds\")\n",
    "\n",
    "CTRL_MUT<-readNformat(\"control_MUT_sorted.new_cbust_result.tsv\")\n",
    "CTRL_REF<-readNformat(\"control_REF_sorted.new_cbust_result.tsv\")\n",
    "dim(CTRL_MUT)\n",
    "dim(CTRL_REF)\n",
    "CTRL_delta<-CTRL_MUT - CTRL_REF\n",
    "saveRDS(CTRL_delta, \"CTRL_delta.Rds\")\n",
    "fdr01_global <- quantile(abs(CTRL_delta), probs=0.999);fdr01_global\n",
    "# 8.135\n",
    "# 3.28\n",
    "fdr1_global <- quantile(abs(CTRL_delta), probs=0.99);fdr1_global\n",
    "# 5.18\n",
    "# 1.129\n",
    "# delta threshold 3\n",
    "nMotifsD3 <- cbind(randomAbs= colSums(abs(CTRL_delta) > 3))\n",
    "nRand <- nrow(CTRL_delta)\n",
    "nMotifsD3 <- cbind(nMotifsD3,\n",
    "                   randomUp=colSums(CTRL_delta > 3),\n",
    "                   randomDw=colSums(CTRL_delta < -3))\n",
    "\n",
    "nSignif <- nrow(ASB_delta)\n",
    "nMotifsD3 <- cbind(nMotifsD3,\n",
    "                   signifAbs=colSums(abs(ASB_delta) > 3),\n",
    "                   signifUp=colSums(ASB_delta > 3),\n",
    "                   signifDw=colSums(ASB_delta < -3))\n",
    "\n",
    "nMotifsD4 <- cbind(randomAbs= colSums(abs(CTRL_delta) > 4))\n",
    "nRand <- nrow(CTRL_delta)\n",
    "nMotifsD4 <- cbind(nMotifsD4,\n",
    "                   randomUp=colSums(CTRL_delta > 4),\n",
    "                   randomDw=colSums(CTRL_delta < -4))\n",
    "\n",
    "nSignif <- nrow(ASB_delta)\n",
    "nMotifsD4 <- cbind(nMotifsD4,\n",
    "                   signifAbs=colSums(abs(ASB_delta) > 4),\n",
    "                   signifUp=colSums(ASB_delta > 4),\n",
    "                   signifDw=colSums(ASB_delta < -4))\n",
    "\n",
    "# Loop for calculating the Fisher exact test for each motif\n",
    "nMotifsD3<-nMotifsD3[rowSums(nMotifsD3)>0,]\n",
    "nMotifsD4<-nMotifsD4[rowSums(nMotifsD4)>0,]\n",
    "\n",
    "# cbust threshold 3\n",
    "fisherTestP_cbust3 <- matrix(0, nrow=nrow(nMotifsD3), ncol=6, dimnames=list(rownames(nMotifsD3),c(\"Abs_gr\",\"Abs_less\",\"Up_gr\",\"Up_less\",\"Dw_gr\",\"Dw_less\")))\n",
    "fisherTestP_cbust4 <- matrix(0, nrow=nrow(nMotifsD4), ncol=6, dimnames=list(rownames(nMotifsD4),c(\"Abs_gr\",\"Abs_less\",\"Up_gr\",\"Up_less\",\"Dw_gr\",\"Dw_less\")))\n",
    "\n",
    "# ABSOLUTE - cbust3\n",
    "for (i in rownames(nMotifsD3))\n",
    "{\n",
    "  nSignifMod <- nMotifsD3[i,\"signifAbs\"]\n",
    "  nRndMod <- nMotifsD3[i,\"randomAbs\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "  fisherTestP_cbust3[i, \"Abs_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust3[i, \"Abs_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "# ABSOLUTE - cbust4\n",
    "for (i in rownames(nMotifsD4))\n",
    "{\n",
    "  nSignifMod <- nMotifsD4[i,\"signifAbs\"]\n",
    "  nRndMod <- nMotifsD4[i,\"randomAbs\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "  fisherTestP_cbust4[i, \"Abs_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust4[i, \"Abs_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "# UP - cbust3\n",
    "for (i in rownames(nMotifsD3))\n",
    "{\n",
    "  nSignifMod <- nMotifsD3[i,\"signifUp\"]\n",
    "  nRndMod <- nMotifsD3[i,\"randomUp\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "\n",
    "  fisherTestP_cbust3[i, \"Up_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust3[i, \"Up_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "# UP - cbust4\n",
    "for (i in rownames(nMotifsD4))\n",
    "{\n",
    "  nSignifMod <- nMotifsD4[i,\"signifUp\"]\n",
    "  nRndMod <- nMotifsD4[i,\"randomUp\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "\n",
    "  fisherTestP_cbust4[i, \"Up_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust4[i, \"Up_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "\n",
    "# DOWN - cbust3\n",
    "for (i in rownames(nMotifsD3))\n",
    "{\n",
    "  nSignifMod <- nMotifsD3[i,\"signifDw\"]\n",
    "  nRndMod <- nMotifsD3[i,\"randomDw\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "\n",
    "  fisherTestP_cbust3[i, \"Dw_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust3[i, \"Dw_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "# DOWN - cbust4\n",
    "for (i in rownames(nMotifsD4))\n",
    "{\n",
    "  nSignifMod <- nMotifsD4[i,\"signifDw\"]\n",
    "  nRndMod <- nMotifsD4[i,\"randomDw\"]\n",
    "  mat <- matrix(c(nSignifMod, nSignif-nSignifMod,\n",
    "                  nRndMod, nRand-nRndMod),\n",
    "                ncol=2, byrow=T) # t() : does not matter\n",
    "\n",
    "  fisherTestP_cbust4[i, \"Dw_less\"] <- fisher.test(mat, alternative = \"less\")$p.value\n",
    "  fisherTestP_cbust4[i, \"Dw_gr\"] <- fisher.test(mat, alternative = \"greater\")$p.value\n",
    "}\n",
    "\n",
    "fisherTestP_cbust3 <- fisherTestP_cbust3[order(fisherTestP_cbust3[,\"Abs_gr\"]),]\n",
    "saveRDS(fisherTestP_cbust3, file=\"fisherTestP_cbust3.Rds\")\n",
    "\n",
    "fisherTestP_cbust4 <- fisherTestP_cbust4[order(fisherTestP_cbust4[,\"Abs_gr\"]),]\n",
    "saveRDS(fisherTestP_cbust4, file=\"fisherTestP_cbust4.Rds\")\n",
    "\n",
    "\n",
    "fisherTest_cbust3_Padj <- apply(fisherTestP_cbust3, 2, p.adjust, method=\"fdr\")\n",
    "fisherTest_cbust4_Padj <- apply(fisherTestP_cbust4, 2, p.adjust, method=\"fdr\")\n",
    "\n",
    "saveRDS(fisherTest_cbust3_Padj, file=\"fisherTest_cbust3_Padj.Rds\")\n",
    "saveRDS(fisherTest_cbust4_Padj, file=\"fisherTest_cbust4_Padj.Rds\")\n",
    "\n",
    "fisherTest_cbust3_Padj <- cbind(fisherTest_cbust3_Padj, nSignif=rowSums(fisherTest_cbust3_Padj < 0.01), nSignifNoAdj=rowSums(fisherTestP_cbust3 < 0.01))\n",
    "fisherTestPadj_cbust3_signif <- fisherTest_cbust3_Padj[fisherTest_cbust3_Padj[,\"nSignifNoAdj\"]>0,, drop=FALSE]\n",
    "\n",
    "fisherTest_cbust4_Padj <- cbind(fisherTest_cbust4_Padj, nSignif=rowSums(fisherTest_cbust4_Padj < 0.01), nSignifNoAdj=rowSums(fisherTestP_cbust4 < 0.01))\n",
    "fisherTestPadj_cbust4_signif <- fisherTest_cbust4_Padj[fisherTest_cbust4_Padj[,\"nSignifNoAdj\"]>0,, drop=FALSE]\n",
    "\n",
    "\n",
    "ASB_delta_mean <- colMeans(ASB_delta)\n",
    "maxDelta <- max(abs(ASB_delta_mean))\n",
    "\n",
    "png(\"fisherTest_new_cbust_3.png\", width=900, height=900, res=96, type=\"cairo\")\n",
    "plot(cbind(deltaMean=ASB_delta_mean[rownames(fisherTest_cbust3_Padj)],\n",
    "           \"-logAdjPval\"=-log(fisherTest_cbust3_Padj[,\"Up_gr\"])), pch=20, cex=1.5,\n",
    "     xlim=c(-maxDelta,maxDelta), main=\"Up_gr\")\n",
    "abline(v=0, col=\"grey\", lty=2)\n",
    "abline(h=-log(0.05), col=\"grey\", lty=3)\n",
    "abline(h=-log(0.01), col=\"darkgrey\", lty=2)\n",
    "dev.off()\n",
    "\n",
    "png(\"fisherTest_new_cbust_3_Abs_gr.png\", width=900, height=900, res=96, type=\"cairo\")\n",
    "plot(cbind(deltaMean=ASB_delta_mean[rownames(fisherTest_cbust3_Padj)],\n",
    "           \"-logAdjPval\"=-log(fisherTest_cbust3_Padj[,\"Abs_gr\"])), pch=20, cex=1.5,\n",
    "     xlim=c(-maxDelta,maxDelta), main=\"Abs_gr\")\n",
    "abline(v=0, col=\"grey\", lty=2)\n",
    "abline(h=-log(0.05), col=\"grey\", lty=3)\n",
    "abline(h=-log(0.01), col=\"darkgrey\", lty=2)\n",
    "dev.off()\n",
    "\n",
    "png(\"fisherTest_new_cbust_4.png\", width=900, height=900, res=96, type=\"cairo\")\n",
    "plot(cbind(deltaMean=ASB_delta_mean[rownames(fisherTest_cbust4_Padj)],\n",
    "           \"-logAdjPval\"=-log(fisherTest_cbust4_Padj[,\"Up_gr\"])), pch=20, cex=1.5,\n",
    "     xlim=c(-maxDelta,maxDelta), main=\"Up_gr\")\n",
    "abline(v=0, col=\"grey\", lty=2)\n",
    "abline(h=-log(0.05), col=\"grey\", lty=3)\n",
    "abline(h=-log(0.01), col=\"darkgrey\", lty=2)\n",
    "dev.off()\n",
    "\n",
    "png(\"fisherTest_new_cbust_4_Abs_gr.png\", width=900, height=900, res=96, type=\"cairo\")\n",
    "plot(cbind(deltaMean=ASB_delta_mean[rownames(fisherTest_cbust4_Padj)],\n",
    "           \"-logAdjPval\"=-log(fisherTest_cbust4_Padj[,\"Abs_gr\"])), pch=20, cex=1.5,\n",
    "     xlim=c(-maxDelta,maxDelta), main=\"Abs_gr\")\n",
    "abline(v=0, col=\"grey\", lty=2)\n",
    "abline(h=-log(0.05), col=\"grey\", lty=3)\n",
    "abline(h=-log(0.01), col=\"darkgrey\", lty=2)\n",
    "dev.off()\n",
    "\n",
    "# Annotation\n",
    "motifAnnotations<-importAnnotations(\"/staging/leuven/stg_00002/lcb/icistarget/data/motif2tf/snapshots/motifs-v8-nr.hgnc-m0.001-o0.0.tbl\")\n",
    "\n",
    "deltaMotifStats_cbust3 <- data.frame(fisherTestPadj_cbust3_signif)\n",
    "annots <- t(sapply(rownames(deltaMotifStats_cbust3), function(x)\n",
    "  c(directOrtology=paste0(unlist(motifAnnotations[motif==x & (directAnnotation | inferred_Orthology),\"TF\"]), collapse=\" ; \"),\n",
    "    inferred_simil=paste0(unlist(motifAnnotations[motif==x & !(directAnnotation | inferred_Orthology),\"TF\"]), collapse=\" ; \"))))\n",
    "deltaMotifStats_cbust3 <- data.frame(delta=signif(ASB_delta_mean[rownames(deltaMotifStats_cbust3)],3), deltaMotifStats_cbust3, annots)\n",
    "deltaMotifStats_cbust3$Abs_gr <- signif(deltaMotifStats_cbust3$Abs_gr,3)\n",
    "deltaMotifStats_cbust3$Abs_less <- signif(deltaMotifStats_cbust3$Abs_less,3)\n",
    "deltaMotifStats_cbust3$Up_gr <- signif(deltaMotifStats_cbust3$Up_gr,3)\n",
    "deltaMotifStats_cbust3$Up_less <- signif(deltaMotifStats_cbust3$Up_less,3)\n",
    "deltaMotifStats_cbust3$Dw_gr <- signif(deltaMotifStats_cbust3$Dw_gr,3)\n",
    "deltaMotifStats_cbust3$Dw_less <- signif(deltaMotifStats_cbust3$Dw_less,3)\n",
    "selCols <- c(\"delta\",\"Abs_gr\", \"Abs_less\", \"Up_gr\", \"Up_less\", \"Dw_gr\", \"Dw_less\", \"directOrtology\", \"inferred_simil\")\n",
    "table2show <- deltaMotifStats_cbust3[,selCols, drop=FALSE]\n",
    "write.table(file=\"motifs_signif_fisher_cbust3.tsv\",table2show,sep=\"\\t\",quote=F)\n",
    "save(table2show, file=\"motifs_signif_fisher_cbust3.RData\")\n",
    "y <- DT::datatable(RcisTarget::addLogo(data.frame(motif=rownames(table2show), table2show)), escape=FALSE)\n",
    "DT::saveWidget(y, 'motifs_signif_fisher_cbust3.html', selfcontained=F)\n",
    "rm(table2show)\n",
    "\n",
    "deltaMotifStats_cbust4 <- data.frame(fisherTestPadj_cbust4_signif)\n",
    "annots <- t(sapply(rownames(deltaMotifStats_cbust4), function(x)\n",
    "  c(directOrtology=paste0(unlist(motifAnnotations[motif==x & (directAnnotation | inferred_Orthology),\"TF\"]), collapse=\" ; \"),\n",
    "    inferred_simil=paste0(unlist(motifAnnotations[motif==x & !(directAnnotation | inferred_Orthology),\"TF\"]), collapse=\" ; \"))))\n",
    "deltaMotifStats_cbust4 <- data.frame(delta=signif(ASB_delta_mean[rownames(deltaMotifStats_cbust4)],3), deltaMotifStats_cbust4, annots)\n",
    "deltaMotifStats_cbust4$Abs_gr <- signif(deltaMotifStats_cbust4$Abs_gr,3)\n",
    "deltaMotifStats_cbust4$Abs_less <- signif(deltaMotifStats_cbust4$Abs_less,3)\n",
    "deltaMotifStats_cbust4$Up_gr <- signif(deltaMotifStats_cbust4$Up_gr,3)\n",
    "deltaMotifStats_cbust4$Up_less <- signif(deltaMotifStats_cbust4$Up_less,3)\n",
    "deltaMotifStats_cbust4$Dw_gr <- signif(deltaMotifStats_cbust4$Dw_gr,3)\n",
    "deltaMotifStats_cbust4$Dw_less <- signif(deltaMotifStats_cbust4$Dw_less,3)\n",
    "selCols <- c(\"delta\",\"Abs_gr\", \"Abs_less\", \"Up_gr\", \"Up_less\", \"Dw_gr\", \"Dw_less\", \"directOrtology\", \"inferred_simil\")\n",
    "table2show <- deltaMotifStats_cbust4[,selCols, drop=FALSE]\n",
    "write.table(file=\"motifs_signif_fisher_cbust4.tsv\",table2show,sep=\"\\t\",quote=F)\n",
    "save(table2show, file=\"motifs_signif_fisher_cbust4.RData\")\n",
    "y <- DT::datatable(RcisTarget::addLogo(data.frame(motif=rownames(table2show), table2show)), escape=FALSE)\n",
    "DT::saveWidget(y, 'motifs_signif_fisher_cbust4.html', selfcontained=F)\n",
    "\n",
    "\n",
    "table2show<-read.delim(file=\"motifs_signif_fisher_cbust3.tsv\",header=T)\n",
    "ASB_delta<-readRDS(\"ASB_delta.Rds\")\n",
    "\n",
    "AP1_motifs<-table2show[union(grep(\"JUN|FOS|BACH\",table2show$directOrtology),grep(\"JUN|FOS|BACH\",table2show$inferred_simil)),]\n",
    "AP1_motifs<-subset(AP1_motifs,Up_gr<=0.05)\n",
    "SOX_motifs<-table2show[union(grep(\"SOX\",table2show$directOrtology),grep(\"SOX\",table2show$inferred_simil)),]\n",
    "SOX_motifs<-subset(SOX_motifs,Up_gr<=0.05)\n",
    "CTCF_motifs<-table2show[union(grep(\"CTCF\",table2show$directOrtology),grep(\"CTCF\",table2show$inferred_simil)),]\n",
    "CTCF_motifs<-subset(CTCF_motifs,Up_gr<=0.05)\n",
    "TFAP_motifs<-table2show[union(grep(\"TFAP\",table2show$directOrtology),grep(\"TFAP\",table2show$inferred_simil)),]\n",
    "TFAP_motifs<-subset(TFAP_motifs,Up_gr<=0.05)\n",
    "ETS_motifs<-table2show[union(grep(\"ETV\",table2show$directOrtology),grep(\"ETV\",table2show$inferred_simil)),]\n",
    "ETS_motifs<-subset(ETS_motifs,Up_gr<=0.05)\n",
    "MYB_motifs<-table2show[union(grep(\"MYB\",table2show$directOrtology),grep(\"MYB\",table2show$inferred_simil)),]\n",
    "MYB_motifs<-subset(MYB_motifs,Up_gr<=0.05)\n",
    "FOXP3_motifs<-table2show[union(grep(\"FOXP3\",table2show$directOrtology),grep(\"FOXP3\",table2show$inferred_simil)),]\n",
    "FOXP3_motifs<-subset(FOXP3_motifs,Up_gr<=0.05)\n",
    "CREB3_motifs<-table2show[union(grep(\"CREB3\",table2show$directOrtology),grep(\"CREB3\",table2show$inferred_simil)),]\n",
    "CREB3_motifs<-subset(CREB3_motifs,Up_gr<=0.05)\n",
    "RUNX_motifs<-table2show[union(grep(\"RUNX3|RUNX2\",table2show$directOrtology),grep(\"RUNX3|RUNX2\",table2show$inferred_simil)),]\n",
    "RUNX_motifs<-subset(RUNX_motifs,Up_gr<=0.05)\n",
    "HOX_motifs<-table2show[union(grep(\"HOX\",table2show$directOrtology),grep(\"HOX\",table2show$inferred_simil)),]\n",
    "HOX_motifs<-subset(HOX_motifs,Up_gr<=0.05)\n",
    "TEAD_motifs<-table2show[union(grep(\"TEAD1|TEAD3|TEAD4\",table2show$directOrtology),grep(\"TEAD1|TEAD3|TEAD4\",table2show$inferred_simil)),]\n",
    "TEAD_motifs<-subset(TEAD_motifs,Up_gr<=0.05)\n",
    "EBOX_motifs<-table2show[union(grep(\"USF|MITF|ARNT|CLOCK|MYC|NPAS2|MNT|HES5|HES7\",table2show$directOrtology),grep(\"USF|MITF|ARNT|CLOCK|MYC|NPAS2|MNT|HES5|HES7\",table2show$inferred_simil)),]\n",
    "EBOX_motifs<-subset(EBOX_motifs,Up_gr<=0.05)\n",
    "SP_motifs<-table2show[union(grep(\"SP2|SP3|SP3|WT1|EGR1|ZNF740|MECOM|ZNF148|ZNF219|CHURC1|OVOL2|IKZF3|MAZ|KLF4\",table2show$directOrtology),grep(\"SP2|SP3|SP3|WT1|EGR1|ZNF740|MECOM|ZNF148|ZNF219|CHURC1|OVOL2|IKZF3|MAZ|KLF4\",table2show$inferred_simil)),]\n",
    "SP_motifs<-subset(SP_motifs,Up_gr<=0.05)\n",
    "\n",
    "if (!all(is.na(AP1_motifs))){AP1_motifs$ann<-\"AP1\"}\n",
    "if (!all(is.na(SOX_motifs))){SOX_motifs$ann<-\"SOX\"}\n",
    "if (!all(is.na(CTCF_motifs))){CTCF_motifs$ann<-\"CTCF\"}\n",
    "if (!all(is.na(TFAP_motifs))){TFAP_motifs$ann<-\"TFAP\"}\n",
    "if (!all(is.na(ETS_motifs))){ETS_motifs$ann<-\"ETS\"}\n",
    "if (!all(is.na(MYB_motifs))){MYB_motifs$ann<-\"MYB\"}\n",
    "if (!all(is.na(FOXP3_motifs))){FOXP3_motifs$ann<-\"FOXP3\"}\n",
    "if (!all(is.na(CREB3_motifs))){CREB3_motifs$ann<-\"CREB3\"}\n",
    "if (!all(is.na(RUNX_motifs))){RUNX_motifs$ann<-\"RUNX\"}\n",
    "if (!all(is.na(HOX_motifs))){HOX_motifs$ann<-\"HOX\"}\n",
    "if (!all(is.na(TEAD_motifs))){TEAD_motifs$ann<-\"TEAD\"}\n",
    "if (!all(is.na(EBOX_motifs))){EBOX_motifs$ann<-\"EBOX\"}\n",
    "if (!all(is.na(SP_motifs))){SP_motifs$ann<-\"SP\"}\n",
    "\n",
    "colors<-c(\"red2\", \"darkgreen\", \"violet\", \"yellow3\", \"deepskyblue\",\"plum1\", \"limegreen\", \"olivedrab3\", \"snow4\",\"darkorange\",\"deeppink\",\"coral4\",\"tomato\",\"royalblue3\")\n",
    "names(colors)<-c(\"AP1\", \"CTCF\",\"CREB3\",\"EBOX\",\"ETS\",\"FOXP3\",\"HOX\",\"MYB\",\"Other\", \"RUNX\",\"SOX\",\"SP\",\"TEAD\",\"TFAP\")\n",
    "\n",
    "motif_ann<-rbind(AP1_motifs, SOX_motifs, CTCF_motifs, TFAP_motifs, ETS_motifs,MYB_motifs, FOXP3_motifs, CREB3_motifs, RUNX_motifs,HOX_motifs, TEAD_motifs,EBOX_motifs, SP_motifs)\n",
    "motif_ann<-motif_ann[,\"ann\",drop=F]\n",
    "ASB_delta_sig_motifs<-ASB_delta[,rownames(subset(table2show,Up_gr<=0.05))]\n",
    "ASB_delta_sig_motifs_binarized<-apply(ASB_delta_sig_motifs,1,function(x){ifelse(x>=3,1,0)})\n",
    "\n",
    "a<-as.data.frame(rowSums(ASB_delta_sig_motifs_binarized))\n",
    "colnames(a)<-\"rowsum\"\n",
    "\n",
    "table2plot<-merge(table2show,a,by.x=\"row.names\",by.y=\"row.names\",all.x=T)\n",
    "table2plot<-merge(table2plot, motif_ann, by.x=\"Row.names\",by.y=\"row.names\", all.x=T)\n",
    "\n",
    "table2plot$rowsum[is.na(table2plot$rowsum)]<- 0.1\n",
    "table2plot$ann[is.na(table2plot$ann)]<-\"Other\"\n",
    "\n",
    "ASB_delta_mean <- colMeans(ASB_delta)\n",
    "maxDelta <- max(abs(ASB_delta_mean))\n",
    "\n",
    "\n",
    "png(file=\"fisherTest_new_cbust_3_ggplot.png\",width=1200, height=1200,res=120, type=\"cairo\")\n",
    "ggplot(table2plot, aes(delta, -log(Up_gr), color=ann, size=rowsum)) +\n",
    "xlim(-maxDelta,maxDelta) +\n",
    "geom_point(alpha=0.9) +\n",
    "scale_colour_manual(values = colors[levels(as.factor(table2plot$ann))] ) + geom_hline(yintercept=-log(0.05),linetype=2,col=\"grey\") +\n",
    "geom_hline(yintercept=-log(0.01),linetype=2,col=\"grey\") +\n",
    "geom_vline(xintercept=0,linetype=2,col=\"grey\") + theme_bw() +\n",
    "theme(legend.title=element_blank(),\n",
    "    legend.text=element_text(size=12),\n",
    "    axis.text.x=element_text(size=12),\n",
    "    axis.text.y=element_text(size=12),\n",
    "    title=element_text(size=16)) + guides(color = guide_legend(override.aes = list(size=6))) +\n",
    "    annotate(\"text\", label=paste(paste(dim(ASB_MUT)[1],\"test\", sep=\" \"), paste(dim(CTRL_MUT)[1],\"control\",sep=\" \"),sep=\",\"), x=-maxDelta+0.05, y=0 )\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
