{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2. Mapping functional genomics data to personalized genomes "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add detailed explanation here"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# load necessary modules\n",
    "module load Bowtie2/2.2.6-foss-2014a\n",
    "module load SAMtools/1.8-foss-2014a\n",
    "module load Picard/1.134-Java-1.7.0_51\n",
    "module load bedtools/20181008-foss-2014a\n",
    "module load Kent/20180816\n",
    "module load Anaconda/5.1.0-Python-2.7\n",
    "source activate macs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# provide the sample name and the fastq file to be mapped here \n",
    "sample=\"MM031\"\n",
    "atac_fastq=\"/staging/leuven/stg_00002/lcb/zkalender/melanoma_ATAC/00.FASTQ/MM031_BL/OmniATAC__MM031__untreated_S3_R2_001.fastq.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create necessary folders \n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC\n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map\n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map\n",
    "\n",
    "cd /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map to the reference genome\n",
    "# map\n",
    "bowtie2 -p 20 --very-sensitive \\\n",
    "-x /staging/leuven/stg_00002/lcb/zkalender/resources/refdata-GRCh38-2.1.0/fasta/genome \\\n",
    "-U ${atac_fastq}  \\\n",
    " | samtools view -b - | samtools sort -@ 20 -l 1 -O bam \\\n",
    "-o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.bam\n",
    "\n",
    "samtools flagstat \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.bam \\\n",
    "> /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.bam.flagstat\n",
    "\n",
    "# duplicate removal\n",
    "picard MarkDuplicates \\\n",
    "I=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.bam \\\n",
    "O=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.bam \\\n",
    "CREATE_INDEX=true \\\n",
    "VALIDATION_STRINGENCY=SILENT \\\n",
    "REMOVE_DUPLICATES=TRUE \\\n",
    "M=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.dedup.metrics\n",
    "\n",
    "samtools view -@ 10 -q 2 -b /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.bam  |\\\n",
    " samtools sort - -@ 10 -O bam -l 9 -T ${VSC_SCRATCH}/tmp -o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.Q2.bam\n",
    "\n",
    "samtools flagstat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.Q2.bam > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.Q2.flagstat\n",
    "\n",
    "# peak calling\n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks\n",
    "\n",
    "macs2 callpeak \\\n",
    "-t /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/${sample}.ATAC.REF.sorted.dedup.Q2.bam \\\n",
    " -q 0.05 \\\n",
    "-n /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks \\\n",
    "--nomodel --shift -75 --extsize 150 --keep-dup all --call-summits \\\n",
    "-f BAM\n",
    "\n",
    "intersectBed -a /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.bed \\\n",
    "-b /staging/leuven/stg_00002/lcb/zkalender/resources/hg38/hg38.canonical_chromosomes.bed \\\n",
    "-wa -nonamecheck | sort -u | sortBed -i stdin > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits_filt.bed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map to HAP1 \n",
    "bowtie2 -p 20 --very-sensitive \\\n",
    "-x /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/liftover_proc/bowtie2_index/${sample}.hap1.fa \\\n",
    "-U ${atac_fastq}  \\\n",
    " | samtools view -b - | samtools sort -@ 20 -l 1 -O bam \\\n",
    "-o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.bam\n",
    "\n",
    "samtools flagstat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.bam > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.bam.flagstat\n",
    "\n",
    "# duplicate removal\n",
    "picard MarkDuplicates \\\n",
    "I=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.bam \\\n",
    "O=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.bam \\\n",
    "CREATE_INDEX=true \\\n",
    "VALIDATION_STRINGENCY=SILENT \\\n",
    "REMOVE_DUPLICATES=TRUE \\\n",
    "M=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.dedup.metrics\n",
    "\n",
    "# quality filter\n",
    "samtools view -@ 10 -q 2 -b /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.bam  |\\\n",
    " samtools sort - -@ 10 -O bam -l 9 -T ${VSC_SCRATCH}/tmp -o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.Q2.bam\n",
    "\n",
    "samtools flagstat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.Q2.bam > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.Q2.flagstat\n",
    "\n",
    "# call peaks on reads mapped to HAP1\n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks\n",
    "macs2 callpeak \\\n",
    "-t /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/${sample}.ATAC.HAP1_ind_map.sorted.dedup.Q2.bam \\\n",
    " -q 0.05 \\\n",
    "-n /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks \\\n",
    "--nomodel --shift -75 --extsize 150 --keep-dup all --call-summits \\\n",
    "-f BAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map to HAP2 \n",
    "bowtie2 -p 20 --very-sensitive \\\n",
    "-x /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/liftover_proc/bowtie2_index/${sample}.hap2.fa \\\n",
    "-U ${atac_fastq}  \\\n",
    " | samtools view -b - | samtools sort -@ 20 -l 1 -O bam \\\n",
    "-o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.bam\n",
    "\n",
    "samtools flagstat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.bam > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.bam.flagstat\n",
    "\n",
    "# duplicate removal\n",
    "picard MarkDuplicates \\\n",
    "I=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.bam \\\n",
    "O=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.bam \\\n",
    "CREATE_INDEX=true \\\n",
    "VALIDATION_STRINGENCY=SILENT \\\n",
    "REMOVE_DUPLICATES=TRUE \\\n",
    "M=/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.dedup.metrics\n",
    "\n",
    "# quality filter\n",
    "samtools view -@ 10 -q 2 -b /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.bam  |\\\n",
    " samtools sort - -@ 10 -O bam -l 9 -T ${VSC_SCRATCH}/tmp \\\n",
    " -o /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.Q2.bam\n",
    "\n",
    "samtools flagstat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.Q2.bam > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.Q2.flagstat\n",
    "\n",
    "# call peaks on reads mapped to HAP2\n",
    "mkdir -p /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks\n",
    "macs2 callpeak \\\n",
    "-t /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/${sample}.ATAC.HAP2_ind_map.sorted.dedup.Q2.bam \\\n",
    " -q 0.05 \\\n",
    "-n /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks \\\n",
    "--nomodel --shift -75 --extsize 150 --keep-dup all --call-summits \\\n",
    "-f BAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Consolidate peaks\n",
    "\n",
    "liftOver /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/liftover_proc/${sample}.hap1_to_hg38.over.chain \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.unmapp\n",
    "\n",
    "liftOver /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/liftover_proc/${sample}.hap2_to_hg38.over.chain \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.unmapp\n",
    "\n",
    "\n",
    "# extend to 500 bp & filter for blacklisted regions\n",
    "# HAP1\n",
    "slopBed -i \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.bed \\\n",
    "-b 250 -g /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.canonical.chrom.sizes | \\\n",
    "subtractBed -a stdin -b /staging/leuven/stg_00002/lcb/zkalender/resources/hg38/peakPass60Perc_sorted.bed -A -wa \\\n",
    "> /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended.bed\n",
    "\n",
    "# HAP2\n",
    "slopBed -i  /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.bed \\\n",
    "-b 250 -g /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.canonical.chrom.sizes | \\\n",
    "subtractBed -a stdin -b /staging/leuven/stg_00002/lcb/zkalender/resources/hg38/peakPass60Perc_sorted.bed -A -wa > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended.bed\n",
    "\n",
    "# REF\n",
    "slopBed -i  /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits_filt.bed \\\n",
    "-b 250 -g /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.canonical.chrom.sizes | \\\n",
    "subtractBed -a stdin -b /staging/leuven/stg_00002/lcb/zkalender/resources/hg38/peakPass60Perc_sorted.bed -A -wa > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended.bed\n",
    "\n",
    "cat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended.bed | \\\n",
    "tr '/' '\\t' | cut -f1-3,15,16 > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended.renamed.bed\n",
    "\n",
    "cat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended.bed | \\\n",
    "tr '/' '\\t' | cut -f1-3,15,16 > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended.renamed.bed\n",
    "\n",
    "cat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended.bed | \\\n",
    "tr '/' '\\t' | cut -f1-3,15,16 > \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended.renamed.bed\n",
    "\n",
    "# normalize peak scores\n",
    "module load R/3.4.1-foss-2014a-noX\n",
    "Rscript ~/lcb/zkalender/software/src_zkalender/CDK7_Menin/normalize_fixed_width_peaks.R \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended.renamed.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended_normalized.bed\n",
    "\n",
    "Rscript ~/lcb/zkalender/software/src_zkalender/CDK7_Menin/normalize_fixed_width_peaks.R \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended.renamed.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended_normalized.bed\n",
    "\n",
    "Rscript ~/lcb/zkalender/software/src_zkalender/CDK7_Menin/normalize_fixed_width_peaks.R \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended.renamed.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended_normalized.bed\n",
    "\n",
    "\n",
    "module load gawk/4.1.3\n",
    "module load bedtools/20181008-foss-2014a\n",
    "\n",
    "cat /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP1_ind_map/q005_peaks/${sample}.ATAC.HAP1_ind_map.q_005.peaks_summits.hg38.extended_normalized.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/HAP2_ind_map/q005_peaks/${sample}.ATAC.HAP2_ind_map.q_005.peaks_summits.hg38.extended_normalized.bed \\\n",
    "/staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/REF/q005_peaks/${sample}.ATAC.REF.q_005.peaks_summits.extended_normalized.bed | \\\n",
    "/staging/leuven/stg_00002/lcb/ghuls/software/iterative_peak_filtering/iterative_peak_filtering.sh - /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/${sample}/AS_ATAC/${sample}_REF_HAP1_HAP2_peaks.merged_and_filtered.bed \\\n",
    "/staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.canonical.chrom.sizes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
