{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Step 4. Scoring ASB and control events with cluster-buster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this step, we evaluate the impact of variants on transcription factor motif binding by evaluating cluster-buster scores of ASCAVs over non-ASCAVs. We perform this analysis in peak-wise and not variant-wise. Thus, we start by filtering peaks that contain discordant ASCAVs (ie. the peak has multiple ASCAVs with different winning haplotypes). Then the remaining peaks are lifted over to HAP1 or HAP2 (depending on the winning haplotype) and sequences were extracted from corresponding haplotype fastas. Next, we select control peaks that are not predicted as ASCAVs from the BaalCHIP output, again filtering for discordant variants. Peak positions are lifted over to HAP1 or HAP2, and sequences were extracted as in previous step. For each peak, the sequences from the reference genome were extracted as well. And finally, all sequences are scored with cluster-buster using a motif collection of 22k motifs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load necessary modules\n",
    "module load Kent_tools/20181218-linux.x86_64\n",
    "module load BEDTools/2.27.1-GCCcore-6.4.0\n",
    "module load Cluster-Buster/20180705-foss-2018a\n",
    "module load Python/3.7.0-foss-2018a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample and directories\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### extract test SNVs\n",
    "mkdir -p motif_analysis\n",
    "\n",
    "grep -w -f <(awk '{ if ($7==\"TRUE\") { print $1 } }' Bayes_report.txt) ../m_min6.txt > motif_analysis/ASB_peaks_and_muts\n",
    "\n",
    "echo `cut -f 1 motif_analysis/ASB_peaks_and_muts | sort -u | wc -l` test variants in `cut -f 15 motif_analysis/ASB_peaks_and_muts | sort -u | wc -l` peaks\n",
    "\n",
    "### Extract fasta for peaks with ASB associated SNVs\n",
    "# First look for peaks that have multiple hits\n",
    "# For the first edition, exclude peaks that have singificant alleles from both haplotypes\n",
    "# in the second edition, check if they are all \"winning\" genotypes - if both are winning exclude the peak, if one is winning and the other is losing\n",
    "# then exclude the peak twice - once with each haplotype\n",
    "if [[ $(cut -f 10-15 motif_analysis/ASB_peaks_and_muts | sort -u | cut -f 2 | sort | uniq -d | wc -l) > 0 ]] ; then\n",
    "    echo finding double hit peaks\n",
    "    cut -f 10,15 motif_analysis/ASB_peaks_and_muts | sort -u | cut -f 2 | sort | uniq -d > motif_analysis/tmp_peaks_with_multiple_hits\n",
    "    cat motif_analysis/ASB_peaks_and_muts | awk '{ print $24\"\\t\"$15 }' | cut -f 2 | sort | uniq -d >> motif_analysis/tmp_peaks_with_multiple_hits\n",
    "    echo there are `cat motif_analysis/tmp_peaks_with_multiple_hits | wc -l` peaks with multiple hits\n",
    "\n",
    "    cat motif_analysis/tmp_peaks_with_multiple_hits \\\n",
    "      | while read peak ; do\n",
    "            echo -n ${peak}\" \"; cat motif_analysis/ASB_peaks_and_muts | awk '$15==\"'${peak}'\"' | cut -f 24 | sort -u | tr '\\n' '|' ; echo;\n",
    "        done \\\n",
    "      | awk '{ if ($2==\"HAP1_LOSES|HAP2_WINS|\" || $2==\"HAP1_WINS|HAP2_LOSES|\") { print $1 } }' \\\n",
    "      > motif_analysis/double_hit_peaks\n",
    "\n",
    "    cat motif_analysis/tmp_peaks_with_multiple_hits \\\n",
    "      | while read peak ; do\n",
    "            echo -n ${peak}\" \"; cat motif_analysis/ASB_peaks_and_muts | awk '$15==\"'${peak}'\"' | cut -f 24 | sort -u | tr '\\n' '|' ; echo;\n",
    "        done \\\n",
    "      | awk '{ if ($2!=\"HAP1_LOSES|HAP2_WINS|\" && $2!=\"HAP1_WINS|HAP2_LOSES|\") { print $1 } }' \\\n",
    "      > motif_analysis/inconsistent_double_hit_peaks\n",
    "\n",
    "    cat motif_analysis/inconsistent_double_hit_peaks \\\n",
    "      | while read peak ; do\n",
    "            grep -w ${peak} motif_analysis/ASB_peaks_and_muts | cut -f 12-15 | sort -u;\n",
    "        done \\\n",
    "      > motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n",
    "    echo `cat motif_analysis/double_hit_peaks | wc -l` peaks have consistent hits\n",
    "else\n",
    "    echo there are no peaks with inconsistent hits;\n",
    "fi\n",
    "\n",
    "# extract peak positions for HAP1 & HAP2 specific peaks (excluding inconsistent peaks)\n",
    "cat motif_analysis/ASB_peaks_and_muts | awk -v 'OFS=\\t' '{ if ($10==\"1|0\") { print $12, $13, $14, $15\"|\"$24 } }' | sort -u | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa > motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed\n",
    "cat motif_analysis/ASB_peaks_and_muts | awk -v 'OFS=\\t' '{ if ($10==\"1|0\") { print $12, $13, $14, $15\"|\"$24 } }' | sort -u | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa > motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed\n",
    "#  cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed | cut -f 4 | tr '|' '\\t' | cut -f 1 | sort -u | wc -l\n",
    "# this number + number of inconsistent peaks give the total number of inital peaks (as expected)\n",
    "\n",
    "# hg38 peaks to hap1\n",
    "liftOver motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed \\\n",
    "    ../${sample}.refTOhap1.chain \\\n",
    "    motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed \\\n",
    "    motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | wc -l) > 0 ]] ; then\n",
    "    echo `cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP1, exclude and redo liftover\n",
    "    cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | grep -v '^#' | cut -f 1-3 | sort -u > motif_analysis/to_skip\n",
    "    subtractBed -a motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP1.bed\n",
    "    liftOver motif_analysis/tmp.HAP1.bed \\\n",
    "        ../${sample}.refTOhap1.chain \\\n",
    "        motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed \\\n",
    "        motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped\n",
    "    echo `cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP1\n",
    "    rm motif_analysis/to_skip motif_analysis/tmp.HAP1.bed\n",
    "fi\n",
    "\n",
    "\n",
    "# hg38 peaks to hap2\n",
    "liftOver motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed \\\n",
    "    ../${sample}.refTOhap2.chain \\\n",
    "    motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "    motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | wc -l) > 0 ]] ; then\n",
    "    echo `cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP2, exclude and redo liftover\n",
    "    cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | grep -v '^#' | cut -f 1-3 | sort -u > motif_analysis/to_skip\n",
    "    subtractBed -a motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP2.bed\n",
    "    liftOver motif_analysis/tmp.HAP2.bed \\\n",
    "        ../${sample}.refTOhap2.chain \\\n",
    "        motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "        motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped\n",
    "    echo `cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP2\n",
    "    rm motif_analysis/to_skip motif_analysis/tmp.HAP2.bed\n",
    "fi\n",
    "\n",
    "# extract fasta files from hap1 and hap2\n",
    "mkdir -p motif_analysis/ASB_mut_fasta\n",
    "\n",
    "#hap1\n",
    "fastaFromBed \\\n",
    "    -fi  ../${sample}.hap1.fa \\\n",
    "    -bed motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed \\\n",
    "    -fo motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP1_peaks.HAP1.fa \\\n",
    "    -name\n",
    "\n",
    "# hap2\n",
    "fastaFromBed \\\n",
    "    -fi  ../${sample}.hap2.fa \\\n",
    "    -bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "    -fo motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP2_peaks.HAP2.fa \\\n",
    "    -name\n",
    "\n",
    "cat motif_analysis/ASB_mut_fasta/* | paste - - | sort -k 1,1 -t \" \" | tr '\\t' '\\n' > motif_analysis/ASB_MUT_sorted.fa\n",
    "\n",
    "# extract fasta files from REF\n",
    "\n",
    "# filter out any unmapped peak from previous set\n",
    "cat motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP1_peaks.HAP1.fa motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP2_peaks.HAP2.fa | grep '>' | tr -d '>' > tmp\n",
    "awk 'FNR==NR{seen[$1]=1;next} ($4) in seen' tmp <( cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed) > motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed\n",
    "rm tmp\n",
    "\n",
    "mkdir -p motif_analysis/ASB_ref_fasta\n",
    "\n",
    "fastaFromBed \\\n",
    "    -fi /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.fa \\\n",
    "    -bed motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed \\\n",
    "    -fo motif_analysis/ASB_ref_fasta/ASB_peaks_and_muts_ALL_peaks.fa \\\n",
    "    -name\n",
    "\n",
    "cat motif_analysis/ASB_ref_fasta/ASB_peaks_and_muts_ALL_peaks.fa | paste - - | sort -k 1,1 -t \" \" | tr '\\t' '\\n' > motif_analysis/ASB_REF_sorted.fa\n",
    "\n",
    "echo Generated test fasta for `cat motif_analysis/ASB_MUT_sorted.fa | grep '^>' | wc -l` peaks with mutation and `cat motif_analysis/ASB_REF_sorted.fa | grep '^>' | wc -l` w/o mutation\n",
    "\n",
    "rm -f motif_analysis/double_hit_peaks motif_analysis/inconsistent_double_hit_peaks.bed motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control variants\n",
    "grep -w -f <(cat Bayes_report.txt | awk '$7==\"FALSE\"' | cut -f1) ../m_min6.txt > motif_analysis/control_peaks_and_muts\n",
    "\n",
    "# filter out SNVs without a winning haplotype\n",
    "\n",
    "echo `cat motif_analysis/control_peaks_and_muts | cut -f 1 | sort -u | wc -l` control variants in `cat motif_analysis/control_peaks_and_muts | cut -f 15 | sort -u | wc -l` peaks\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts | cut -f 10,15 | sort -u | cut -f 2 | sort | uniq -d | wc -l) > 0 ]]; then\n",
    "    echo finding double hit peaks\n",
    "    cat motif_analysis/control_peaks_and_muts | cut -f 10,15 | sort -u | cut -f2 | sort | uniq -d  > motif_analysis/tmp_peaks_with_multiple_hits\n",
    "    echo there are `cat motif_analysis/tmp_peaks_with_multiple_hits | wc -l` peaks with multiple hits\n",
    "\n",
    "    # check for inconsistent peaks\n",
    "    cat motif_analysis/tmp_peaks_with_multiple_hits \\\n",
    "      | while read peak ; do\n",
    "            echo -n ${peak}\" \"; cat motif_analysis/control_peaks_and_muts | awk '$15==\"'${peak}'\"' | cut -f 24 | sort -u | tr '\\n' '|'; echo;\n",
    "        done \\\n",
    "      | awk '{ if ($2==\"HAP1_LOSES|HAP2_WINS|\" || $2==\"HAP1_WINS|HAP2_LOSES|\") { print $1 } }' \\\n",
    "      > motif_analysis/double_hit_peaks\n",
    "\n",
    "     cat motif_analysis/tmp_peaks_with_multiple_hits \\\n",
    "       | while read peak ; do\n",
    "             echo -n ${peak}\" \"; cat motif_analysis/control_peaks_and_muts | awk '$15==\"'${peak}'\"' | cut -f 24 | sort -u | tr '\\n' '|'; echo;\n",
    "         done \\\n",
    "       | awk '{ if($2!=\"HAP1_LOSES|HAP2_WINS|\" && $2!=\"HAP1_WINS|HAP2_LOSES|\") { print $1 } }' \\\n",
    "       > motif_analysis/inconsistent_double_hit_peaks\n",
    "\n",
    "     cat motif_analysis/inconsistent_double_hit_peaks \\\n",
    "       | while read peak ; do\n",
    "             grep -w ${peak} motif_analysis/control_peaks_and_muts | cut -f 12-15 | sort -u;\n",
    "         done \\\n",
    "       > motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n",
    "    echo `cat motif_analysis/double_hit_peaks | wc -l` peaks have consistent hits\n",
    "else\n",
    "    echo there are no peaks with inconsistent hits;\n",
    "fi\n",
    "\n",
    "# create HAP1 specific and HAP2 specific bed files\n",
    "# filter out inconsistent peaks\n",
    "# filter out peaks in test set\n",
    "cat motif_analysis/control_peaks_and_muts | awk -v 'OFS=\\t' '{ if ($10==\"1|0\") { print $12, $13, $14, $15\"|\"$24 } }' | sort -u \\\n",
    "  | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa \\\n",
    "  | subtractBed -a stdin -b motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed -A -wa \\\n",
    "  > motif_analysis/control_peaks_and_muts_HAP1_peaks.bed\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts | awk -v 'OFS=\\t' '{ if ($10==\"0|1\") { print $12, $13, $14, $15\"|\"$24 } }' | sort -u \\\n",
    "  | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa \\\n",
    "  | subtractBed -a stdin -b motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed -A -wa \\\n",
    "  > motif_analysis/control_peaks_and_muts_HAP2_peaks.bed\n",
    "\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed motif_analysis/control_peaks_and_muts_HAP2_peaks.bed > motif_analysis/control_peaks_and_muts.all_peaks.bed\n",
    "\n",
    "\n",
    "# hg38 peaks to hap1\n",
    "liftOver motif_analysis/control_peaks_and_muts_HAP1_peaks.bed \\\n",
    "    ../${sample}.refTOhap1.chain \\\n",
    "    motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed \\\n",
    "    motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | wc -l) > 0 ]] ; then\n",
    "    echo `cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP1, exclude and redo liftover\n",
    "    cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | grep -v '^#' | cut -f 1-3 | sort -u > motif_analysis/to_skip\n",
    "    subtractBed -a motif_analysis/control_peaks_and_muts_HAP1_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP1.bed\n",
    "    liftOver motif_analysis/tmp.HAP1.bed \\\n",
    "        ../${sample}.refTOhap1.chain \\\n",
    "        motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed \\\n",
    "        motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped\n",
    "        echo `cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP1\n",
    "        rm motif_analysis/to_skip motif_analysis/tmp.HAP1.bed\n",
    "fi\n",
    "\n",
    "# hg38 peaks to hap2\n",
    "liftOver motif_analysis/control_peaks_and_muts_HAP2_peaks.bed \\\n",
    "    ../${sample}.refTOhap2.chain \\\n",
    "    motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed \\\n",
    "    motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | wc -l) > 0 ]] ; then\n",
    "    echo `cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP2, exclude and redo liftover\n",
    "    cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | grep -v '^#' | cut -f 1-3 | sort -u > motif_analysis/to_skip\n",
    "    subtractBed -a motif_analysis/control_peaks_and_muts_HAP2_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP2.bed\n",
    "    liftOver motif_analysis/tmp.HAP2.bed \\\n",
    "        ../${sample}.refTOhap2.chain \\\n",
    "        motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed \\\n",
    "        motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped\n",
    "        echo `cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | grep -v '^#' | wc -l` peak unmapped to HAP2\n",
    "        rm motif_analysis/to_skip motif_analysis/tmp.HAP2.bed\n",
    "fi\n",
    "\n",
    "fastaFromBed \\\n",
    "    -fi ../${sample}.hap1.fa \\\n",
    "    -bed motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed \\\n",
    "    -fo motif_analysis/control_peaks_and_muts_HAP1_peaks.HAP1.fa \\\n",
    "    -name\n",
    "\n",
    "fastaFromBed \\\n",
    "    -fi ../${sample}.hap2.fa \\\n",
    "    -bed motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed \\\n",
    "    -fo motif_analysis/control_peaks_and_muts_HAP2_peaks.HAP2.fa \\\n",
    "    -name\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts_HAP1_peaks.HAP1.fa motif_analysis/control_peaks_and_muts_HAP2_peaks.HAP2.fa \\\n",
    "  | paste - - | sort -k 1 | uniq | sort -k 1,1 -t \" \"| tr '\\t' '\\n' \\\n",
    "  > motif_analysis/control_MUT_sorted.fa\n",
    "\n",
    "# filter out any unmapped peak from previous set\n",
    "cat motif_analysis/control_MUT_sorted.fa | grep '>' | tr -d '>' > tmp\n",
    "awk 'FNR==NR{seen[$1]=1;next} ($4) in seen' tmp motif_analysis/control_peaks_and_muts.all_peaks.bed > motif_analysis/control_peaks_and_muts.all_peaks.filt.bed\n",
    "rm tmp\n",
    "\n",
    "fastaFromBed \\\n",
    "    -fi /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.fa \\\n",
    "    -bed motif_analysis/control_peaks_and_muts.all_peaks.filt.bed \\\n",
    "    -fo motif_analysis/control_REF.fa \\\n",
    "    -name\n",
    "\n",
    "cat motif_analysis/control_REF.fa | paste - - | sort -k 1,1 -t \" \" | tr '\\t' '\\n' \\\n",
    "  > motif_analysis/control_REF_sorted.fa\n",
    "\n",
    "echo Generated control fasta for `cat motif_analysis/control_MUT_sorted.fa | grep '^>' | wc -l` peaks with mutation and `cat motif_analysis/control_REF_sorted.fa | grep '^>' | wc -l` w/o mutation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# motif scoring\n",
    "# ASB_MUT\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py \\\n",
    "    -O tsv \\\n",
    "    -f motif_analysis/ASB_MUT_sorted.fa \\\n",
    "    -M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "    -m ${wgs_folder}/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "    -o motif_analysis/ASB_MUT_sorted.new_cbust_result.tsv \\\n",
    "    -t 30\n",
    "# ASB_REF\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py \\\n",
    "    -O tsv \\\n",
    "    -f motif_analysis/ASB_REF_sorted.fa \\\n",
    "    -M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "    -m ${wgs_folder}/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "    -o motif_analysis/ASB_REF_sorted.new_cbust_result.tsv \\\n",
    "    -t 30\n",
    "# CTRL_MUT\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py \\\n",
    "    -O tsv \\\n",
    "    -f motif_analysis/control_MUT_sorted.fa \\\n",
    "    -M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "    -m ${wgs_folder}/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "    -o motif_analysis/control_MUT_sorted.new_cbust_result.tsv \\\n",
    "    -t 30\n",
    "# CTRL_REF\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py \\\n",
    "    -O tsv \\\n",
    "    -f motif_analysis/control_REF_sorted.fa \\\n",
    "    -M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "    -m ${wgs_folder}/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "    -o motif_analysis/control_REF_sorted.new_cbust_result.tsv \\\n",
    "    -t 30"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SSH r10n1 Bash",
   "language": "bash",
   "name": "rik_ssh_r10n1_bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
