{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Step 4. Scoring ASB and control events with cluster-buster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Add detailed description here "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load necessary modules \n",
    "module load Kent_tools/20181218-linux.x86_64\n",
    "module load BEDTools/2.27.1-GCCcore-6.4.0\n",
    "module load Cluster-Buster/20180705-foss-2018a\n",
    "module load Python/3.7.0-foss-2018a\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample and directories \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### extract test SNVs\n",
    "mkdir -p motif_analysis\n",
    "\n",
    "grep -w -f <(cat Bayes_report.txt | awk '$7==\"TRUE\"' | cut -f1) ../m_min6.txt > motif_analysis/ASB_peaks_and_muts\n",
    "\n",
    "echo `cat motif_analysis/ASB_peaks_and_muts | cut -f1| sort -u | wc -l` test variants in `cat motif_analysis/ASB_peaks_and_muts | cut -f15| sort -u| wc -l` peaks\n",
    "\n",
    "### Extract fasta for peaks with ASB associated SNVs\n",
    "# First look for peaks that have multiple hits\n",
    "# For the first edition, exclude peaks that have singificant alleles from both haplotypes\n",
    "# in the second edition, check if they are all \"winning\" genotypes - if both are winning exclude the peak, if one is winning and the other is losing\n",
    "# then exclude the peak twice - once with each haplotype\n",
    "if [[ $(cat motif_analysis/ASB_peaks_and_muts | cut -f10,15 | sort -u | cut -f2 | sort | uniq -d  |wc -l) > 0 ]];\n",
    "then echo finding double hit peaks\n",
    "cat motif_analysis/ASB_peaks_and_muts | cut -f10,15 | sort -u | cut -f2 | sort | uniq -d  > motif_analysis/tmp_peaks_with_multiple_hits\n",
    "cat motif_analysis/ASB_peaks_and_muts | awk '{print $24\"\\t\"$15}'| cut -f2 | sort | uniq -d >> motif_analysis/tmp_peaks_with_multiple_hits\n",
    "echo there are `cat motif_analysis/tmp_peaks_with_multiple_hits| wc -l` peaks with multiple hits\n",
    "\n",
    "cat motif_analysis/tmp_peaks_with_multiple_hits | \\\n",
    "while read peak; \\\n",
    "do echo -n $peak\" \"; cat motif_analysis/ASB_peaks_and_muts | awk '$15==\"'$peak'\"'| cut -f24| sort -u| tr '\\n' '|' ;echo;  \\\n",
    "done | awk '$2==\"HAP1_LOSES|HAP2_WINS|\" || $2==\"HAP1_WINS|HAP2_LOSES|\"' | awk '{print $1}' > motif_analysis/double_hit_peaks\n",
    "\n",
    "cat motif_analysis/tmp_peaks_with_multiple_hits | \\\n",
    "while read peak; \\\n",
    "do echo -n $peak\" \"; cat motif_analysis/ASB_peaks_and_muts | awk '$15==\"'$peak'\"'| cut -f24| sort -u| tr '\\n' '|' ;echo;  \\\n",
    "done | awk '$2!=\"HAP1_LOSES|HAP2_WINS|\" && $2!=\"HAP1_WINS|HAP2_LOSES|\"' | awk '{print $1}' > motif_analysis/inconsistent_double_hit_peaks\n",
    "\n",
    "cat motif_analysis/inconsistent_double_hit_peaks | while read peak; \\\n",
    "do grep -w $peak motif_analysis/ASB_peaks_and_muts  | cut -f12-15| sort -u; done > motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n",
    "echo `cat motif_analysis/double_hit_peaks| wc -l` peaks have consistent hits\n",
    "else echo there are no peaks with inconsistent hits;\n",
    "fi\n",
    "\n",
    "# extract peak positions for HAP1  & HAP2 specific peaks (excluding inconsisten peaks)\n",
    "cat motif_analysis/ASB_peaks_and_muts | awk '$10==\"1|0\"' | awk '{print $12,$13,$14,$15\"|\"$24}'| tr ' ' '\\t' |sort -u | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa > motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed\n",
    "cat motif_analysis/ASB_peaks_and_muts | awk '$10==\"0|1\"' | awk '{print $12,$13,$14,$15\"|\"$24}'| tr ' ' '\\t' |sort -u | subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa > motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed\n",
    "#  cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed | cut -f4| tr '|' '\\t' | cut -f1| sort -u | wc -l\n",
    "# this number + number of inconsistent peaks give the total number of inital peaks (as expected)\n",
    "\n",
    "# hg38 peaks to hap1\n",
    "liftOver motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed \\\n",
    "../$sample.refTOhap1.chain \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | wc -l) > 0 ]];\n",
    "then\n",
    "echo `cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped | grep -v '^#' |wc -l` peak unmapped to HAP1, exclude and redo liftover\n",
    " cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped  | grep -v '^#'| cut -f1-3 | sort -u > motif_analysis/to_skip\n",
    "subtractBed -a motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP1.bed\n",
    "liftOver motif_analysis/tmp.HAP1.bed \\\n",
    "../$sample.refTOhap1.chain \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped\n",
    "echo `cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed_unmapped| grep -v '^#'| wc -l` peak unmapped to HAP1\n",
    "rm motif_analysis/to_skip motif_analysis/tmp.HAP1.bed\n",
    "fi\n",
    "\n",
    "\n",
    "# hg38 peaks to hap2\n",
    "liftOver motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed \\\n",
    "../$sample.refTOhap2.chain \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | wc -l) > 0 ]];\n",
    "then\n",
    "echo `cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped | grep -v '^#' |wc -l` peak unmapped to HAP2, exclude and redo liftover\n",
    " cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped  | grep -v '^#'| cut -f1-3 | sort -u > motif_analysis/to_skip\n",
    "subtractBed -a motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP2.bed\n",
    "liftOver motif_analysis/tmp.HAP2.bed \\\n",
    "../$sample.refTOhap2.chain \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped\n",
    "echo `cat motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed_unmapped| grep -v '^#'| wc -l` peak unmapped to HAP2\n",
    "rm motif_analysis/to_skip motif_analysis/tmp.HAP2.bed\n",
    "fi\n",
    "\n",
    "# extract fasta files from hap1 and hap2\n",
    "mkdir -p motif_analysis/ASB_mut_fasta\n",
    "\n",
    "#hap1\n",
    "fastaFromBed \\\n",
    "-fi  ../$sample.hap1.fa \\\n",
    "-bed motif_analysis/ASB_peaks_and_muts_HAP1_peaks.HAP1.bed  \\\n",
    "-fo motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP1_peaks.HAP1.fa -name\n",
    "\n",
    "# hap2\n",
    "fastaFromBed \\\n",
    "-fi  ../$sample.hap2.fa \\\n",
    "-bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.HAP2.bed \\\n",
    "-fo motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP2_peaks.HAP2.fa -name\n",
    "\n",
    "cat motif_analysis/ASB_mut_fasta/* | paste - - | sort -k1,1 -t \" \"| tr '\\t' '\\n' > motif_analysis/ASB_MUT_sorted.fa\n",
    "\n",
    "# extract fasta files from REF\n",
    "\n",
    "# filter out any unmapped peak from previous set\n",
    "cat motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP1_peaks.HAP1.fa motif_analysis/ASB_mut_fasta/ASB_peaks_and_muts_HAP2_peaks.HAP2.fa | grep '>' | tr -d '>' > tmp\n",
    "awk 'FNR==NR{seen[$1]=1;next} ($4) in seen' tmp <( cat motif_analysis/ASB_peaks_and_muts_HAP1_peaks.bed motif_analysis/ASB_peaks_and_muts_HAP2_peaks.bed) > motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed\n",
    "rm tmp\n",
    "\n",
    "mkdir -p motif_analysis/ASB_ref_fasta\n",
    "\n",
    "fastaFromBed \\\n",
    "-fi /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.fa \\\n",
    "-bed motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed -fo motif_analysis/ASB_ref_fasta/ASB_peaks_and_muts_ALL_peaks.fa -name\n",
    "\n",
    "cat motif_analysis/ASB_ref_fasta/ASB_peaks_and_muts_ALL_peaks.fa  | paste - - | sort -k1,1 -t \" \"| tr '\\t' '\\n' > motif_analysis/ASB_REF_sorted.fa\n",
    "\n",
    "echo Generated test fasta for `cat motif_analysis/ASB_MUT_sorted.fa | grep '^>'| wc -l` peaks with mutation and `cat motif_analysis/ASB_REF_sorted.fa | grep '^>'| wc -l` w/o mutation\n",
    "\n",
    "rm -f motif_analysis/double_hit_peaks motif_analysis/inconsistent_double_hit_peaks.bed motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control variants \n",
    "grep -w -f <(cat Bayes_report.txt | awk '$7==\"FALSE\"' | cut -f1) ../m_min6.txt > motif_analysis/control_peaks_and_muts\n",
    "\n",
    "# filter out SNVs without a winning haplotype\n",
    "\n",
    "echo `cat motif_analysis/control_peaks_and_muts | cut -f1| sort -u | wc -l` control variants in `cat motif_analysis/control_peaks_and_muts | cut -f15| sort -u| wc -l` peaks\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts | cut -f10,15 | sort -u | cut -f2 | sort | uniq -d  |wc -l) > 0 ]];\n",
    "then echo finding double hit peaks\n",
    "cat motif_analysis/control_peaks_and_muts | cut -f10,15 | sort -u | cut -f2 | sort | uniq -d  > motif_analysis/tmp_peaks_with_multiple_hits\n",
    "echo there are `cat motif_analysis/tmp_peaks_with_multiple_hits| wc -l` peaks with multiple hits\n",
    "\n",
    "# check for inconsistent peaks\n",
    "cat motif_analysis/tmp_peaks_with_multiple_hits | \\\n",
    "while read peak; \\\n",
    "do echo -n $peak\" \"; cat motif_analysis/control_peaks_and_muts | awk '$15==\"'$peak'\"'| cut -f24| sort -u| tr '\\n' '|' ;echo;  \\\n",
    "done | awk '$2==\"HAP1_LOSES|HAP2_WINS|\" || $2==\"HAP1_WINS|HAP2_LOSES|\"' | awk '{print $1}' > motif_analysis/double_hit_peaks\n",
    "\n",
    "cat motif_analysis/tmp_peaks_with_multiple_hits | \\\n",
    "while read peak; \\\n",
    "do echo -n $peak\" \"; cat motif_analysis/control_peaks_and_muts | awk '$15==\"'$peak'\"'| cut -f24| sort -u| tr '\\n' '|' ;echo;  \\\n",
    "done | awk '$2!=\"HAP1_LOSES|HAP2_WINS|\" && $2!=\"HAP1_WINS|HAP2_LOSES|\"' | awk '{print $1}' > motif_analysis/inconsistent_double_hit_peaks\n",
    "\n",
    "cat motif_analysis/inconsistent_double_hit_peaks | while read peak; \\\n",
    "do grep -w $peak motif_analysis/control_peaks_and_muts  | cut -f12-15| sort -u; done > motif_analysis/inconsistent_double_hit_peaks.bed\n",
    "\n",
    "echo `cat motif_analysis/double_hit_peaks| wc -l` peaks have consistent hits\n",
    "else echo there are no peaks with inconsistent hits;\n",
    "fi\n",
    "\n",
    "# create HAP1 specific and HAP2 specific bed files\n",
    "# filter out inconsistent peaks\n",
    "# filter out peaks in test set\n",
    "cat motif_analysis/control_peaks_and_muts | awk '$10==\"1|0\"' | awk '{print $12,$13,$14,$15\"|\"$24}'| tr ' ' '\\t' |sort -u | \\\n",
    "subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa | \\\n",
    "subtractBed -a stdin -b motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed -A -wa > motif_analysis/control_peaks_and_muts_HAP1_peaks.bed\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts | awk '$10==\"0|1\"' | awk '{print $12,$13,$14,$15\"|\"$24}'| tr ' ' '\\t' |sort -u | \\\n",
    "subtractBed -a stdin -b motif_analysis/inconsistent_double_hit_peaks.bed -A -wa | \\\n",
    "subtractBed -a stdin -b motif_analysis/ASB_peaks_and_muts_ALL_peaks.filt.bed -A -wa > motif_analysis/control_peaks_and_muts_HAP2_peaks.bed\n",
    "\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed motif_analysis/control_peaks_and_muts_HAP2_peaks.bed > motif_analysis/control_peaks_and_muts.all_peaks.bed\n",
    "\n",
    "\n",
    "# hg38 peaks to hap1\n",
    "liftOver motif_analysis/control_peaks_and_muts_HAP1_peaks.bed  \\\n",
    "../$sample.refTOhap1.chain \\\n",
    "motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed \\\n",
    "motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | wc -l) > 0 ]];\n",
    "then\n",
    "echo `cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | grep -v '^#' |wc -l` peak unmapped to HAP1, exclude and redo liftover\n",
    "cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped  | grep -v '^#'| cut -f1-3 | sort -u > motif_analysis/to_skip\n",
    "subtractBed -a motif_analysis/control_peaks_and_muts_HAP1_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP1.bed\n",
    "liftOver motif_analysis/tmp.HAP1.bed \\\n",
    "../$sample.refTOhap1.chain \\\n",
    "motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed \\\n",
    "motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped\n",
    "echo `cat motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed_unmapped | grep -v '^#'| wc -l` peak unmapped to HAP1\n",
    "rm motif_analysis/to_skip motif_analysis/tmp.HAP1.bed\n",
    "fi\n",
    "\n",
    "# hg38 peaks to hap2\n",
    "liftOver motif_analysis/control_peaks_and_muts_HAP2_peaks.bed  \\\n",
    "../$sample.refTOhap2.chain \\\n",
    "motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed \\\n",
    "motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped\n",
    "\n",
    "if [[ $(cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | wc -l) > 0 ]];\n",
    "then\n",
    "echo `cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | grep -v '^#' |wc -l` peak unmapped to HAP2, exclude and redo liftover\n",
    "cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped  | grep -v '^#'| cut -f1-3 | sort -u > motif_analysis/to_skip\n",
    "subtractBed -a motif_analysis/control_peaks_and_muts_HAP2_peaks.bed -b motif_analysis/to_skip -A -wa > motif_analysis/tmp.HAP2.bed\n",
    "liftOver motif_analysis/tmp.HAP2.bed \\\n",
    "../$sample.refTOhap2.chain \\\n",
    "motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed \\\n",
    "motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped\n",
    "echo `cat motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed_unmapped | grep -v '^#'| wc -l` peak unmapped to HAP2\n",
    "rm motif_analysis/to_skip motif_analysis/tmp.HAP2.bed\n",
    "fi\n",
    "\n",
    "fastaFromBed \\\n",
    "-fi   ../$sample.hap1.fa \\\n",
    "-bed motif_analysis/control_peaks_and_muts_HAP1_peaks.bed.HAP1.bed  \\\n",
    "-fo motif_analysis/control_peaks_and_muts_HAP1_peaks.HAP1.fa -name\n",
    "\n",
    "fastaFromBed \\\n",
    "-fi   ../$sample.hap2.fa \\\n",
    "-bed motif_analysis/control_peaks_and_muts_HAP2_peaks.bed.HAP2.bed  \\\n",
    "-fo motif_analysis/control_peaks_and_muts_HAP2_peaks.HAP2.fa -name\n",
    "\n",
    "cat motif_analysis/control_peaks_and_muts_HAP1_peaks.HAP1.fa motif_analysis/control_peaks_and_muts_HAP2_peaks.HAP2.fa | \\\n",
    "paste - - | sort -k1 | uniq  | sort -k1,1 -t \" \"| tr '\\t' '\\n'  > motif_analysis/control_MUT_sorted.fa\n",
    "\n",
    "# filter out any unmapped peak from previous set\n",
    "cat motif_analysis/control_MUT_sorted.fa | grep '>' | tr -d '>' > tmp\n",
    "awk 'FNR==NR{seen[$1]=1;next} ($4) in seen' tmp motif_analysis/control_peaks_and_muts.all_peaks.bed > motif_analysis/control_peaks_and_muts.all_peaks.filt.bed\n",
    "rm tmp\n",
    "\n",
    "fastaFromBed \\\n",
    "-fi /staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.fa \\\n",
    "-bed motif_analysis/control_peaks_and_muts.all_peaks.filt.bed -fo motif_analysis/control_REF.fa -name\n",
    "\n",
    "cat motif_analysis/control_REF.fa  | paste - - | sort -k1,1 -t \" \"| tr '\\t' '\\n' > motif_analysis/control_REF_sorted.fa\n",
    "\n",
    "echo Generated control fasta for `cat motif_analysis/control_MUT_sorted.fa | grep '^>'| wc -l` peaks with mutation and `cat motif_analysis/control_REF_sorted.fa | grep '^>'| wc -l` w/o mutation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# motif scoring \n",
    "# ASB_MUT\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py -O tsv -f motif_analysis/ASB_MUT_sorted.fa \\\n",
    "-M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "-m /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "-o motif_analysis/ASB_MUT_sorted.new_cbust_result.tsv -t 30\n",
    "# ASB_REF\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py -O tsv -f motif_analysis/ASB_REF_sorted.fa \\\n",
    "-M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "-m /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "-o motif_analysis/ASB_REF_sorted.new_cbust_result.tsv -t 30\n",
    "# CTRL_MUT\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py -O tsv -f motif_analysis/control_MUT_sorted.fa \\\n",
    "-M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "-m /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "-o motif_analysis/control_MUT_sorted.new_cbust_result.tsv -t 30\n",
    "# CTRL_REF\n",
    "/staging/leuven/stg_00002/software/primescore/src/make_feature_table.py -O tsv -f motif_analysis/control_REF_sorted.fa \\\n",
    "-M /ddn1/vol1/staging/leuven/stg_00002/lcb/icistarget/data/motifCollection/v8/singletons \\\n",
    "-m /staging/leuven/stg_00002/lcb/zkalender/melanoma_WGS/MM057/alleleseq/motif_analysis/singletons_names.txt \\\n",
    "-o motif_analysis/control_REF_sorted.new_cbust_result.tsv -t 30"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SSH r10n1 Bash",
   "language": "bash",
   "name": "rik_ssh_r10n1_bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
